services/configuration_repository/configuration_repository_service.rb

```diff
@@ -0,0 +1,252 @@
+      return @@mutex
+    end
+    # rubocop:enable Style/ClassVars
+
+    def self.fastlane_ci_config
+      return "fastlane-ci-config"
+    end
+
+    attr_reader :client
+
+    attr_accessor :provider_credential
+
+    # Creates a remote repository if it does not already exist, complete with
+    # the expected remote files `user.json` and `projects.json`
+    def create_private_remote_configuration_repo
+      not_implemented(__method__)
+    end
+
+    # Returns `true` if the configuration repository is in proper format:
+    #
+    #   i.   The repository exists
+    #   ii.  The `users.json` file exists and is a JSON array
+    #   iii. The `projects.json` file exists and is a JSON array
+    #
+    # @return [Boolean]
+    def configuration_repository_valid?
+      not_implemented(__method__)
+    end
+
+    # Returns `true` if the remote configuration repository exists
+    #
+    # @return [Boolean]
+    def configuration_repository_exists?
+      not_implemented(__method__)
+    end
+
+    def clone(path: self.ci_repo_root_path, branch: nil)
+      # This triggers the check of an existing repo in the given path,
+      # we recover from the error making the clone and checkout
+      git = Git.open(File.join(path, ConfigurationRepositoryService.fastlane_ci_config))
+      self.watch_for_changes!(git, path) if File.directory?(path)
+      return git
+    rescue ArgumentError
+      git = @code_hosting_service_class.clone(
+        repo_url: FastlaneCI.env.repo_url,
+        provider_credential: self.provider_credential,
+        path: path,
+        branch: branch,
+        name: "fastlane-ci-config"
+      )
+      self.watch_for_changes!(git, path) if File.directory?(path)
+      return git
+    end
+
+    def pull(path: self.ci_repo_root_path, branch: nil)
+      git = Git.open(File.join(path, ConfigurationRepositoryService.fastlane_ci_config))
+      git.fetch
+      # Are we behind remote? If so, pull.
+      return if git.log.between(git.branch.name, git.remote.branch.name).size.zero?
+      git.pull
+    rescue ArgumentError
+      git = self.clone(path: path, branch: branch)
+      return if git.log.between(git.branch.name, git.remote.branch.name).size.zero?
+      git.pull
+    end
+
+    def file_path(file_path)
+      File.join(self.ci_repo_root_path, ConfigurationRepositoryService.fastlane_ci_config, file_path)
+    end
+
+    protected
+
+    # Serializes CI user and its provider credentials to a JSON format.
+    #
+    # After creating the private remote configuration repository, the server
+    # will stop redirecting the user to the `/configuration` page. In the
+    # `/dashboard` page, there's a requirement the current user has a
+    # provider credential , or it will raise an exception. By writing the
+    # provider credentials to the private configuration repo for the CI User,
+    # it prevents this issue
+    #
+    # @return [String]
+    def serialized_users
+      users = [
+        User.new(
```

Comment:
why do we need this user to be created each time? I feel like we only need that as long as we haven't written out the first users.json file with the ci_user

Reply:
This was part of @AndrewMcBurney code, I conserved it, but he might throw some insight about the purpose of it.

Reply:
Yeah, we could use the `ci_user` too. I didn't think about that when I originally wrote it. 👍

Reply:
@AndrewMcBurney as I don't want to mess up everything, could you please make a PR on this? 🙏